container/containerd/handler.go

```diff
@@ -0,0 +1,244 @@
+	name string,
+	machineInfoFactory info.MachineInfoFactory,
+	fsInfo fs.FsInfo,
+	cgroupSubsystems *containerlibcontainer.CgroupSubsystems,
+	inHostNamespace bool,
+	metadataEnvs []string,
+	ignoreMetrics container.MetricSet,
+) (container.ContainerHandler, error) {
+	// Create the cgroup paths.
+	cgroupPaths := make(map[string]string, len(cgroupSubsystems.MountPoints))
+	for key, val := range cgroupSubsystems.MountPoints {
+		cgroupPaths[key] = path.Join(val, name)
+	}
+
+	// Generate the equivalent cgroup manager for this container.
+	cgroupManager := &cgroupfs.Manager{
+		Cgroups: &libcontainerconfigs.Cgroup{
+			Name: name,
+		},
+		Paths: cgroupPaths,
+	}
+
+	id := ContainerNameToContainerdID(name)
+	// We assume that if load fails then the container is not known to containerd.
+	ctx := context.Background()
+	cntr, err := client.LoadContainer(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	var spec specs.Spec
+	if err := json.Unmarshal(cntr.Spec.Value, &spec); err != nil {
+		return nil, err
+	}
+
+	taskPid, err := client.TaskPid(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	rootfs := "/"
+	if !inHostNamespace {
+		rootfs = RootFSStorageDir + cntr.ID + spec.Root.Path
+	}
+
+	handler := &containerdContainerHandler{
+		id:                 id,
+		client:             client,
+		name:               name,
+		machineInfoFactory: machineInfoFactory,
+		cgroupPaths:        cgroupPaths,
+		cgroupManager:      cgroupManager,
+		rootFs:             rootfs,
+		fsInfo:             fsInfo,
+		envs:               make(map[string]string),
+		labels:             make(map[string]string),
+		ignoreMetrics:      ignoreMetrics,
+		pid:                int(taskPid),
+		creationTime:       cntr.CreatedAt,
+	}
+	// Add the name and bare ID as aliases of the container.
+	handler.labels = cntr.Labels
+	handler.image = cntr.Image
+	handler.aliases = []string{id, name}
+	for _, envVar := range spec.Process.Env {
+		if envVar != "" {
+			splits := strings.SplitN(envVar, "=", 2)
+			if len(splits) == 2 {
+				handler.envs[splits[0]] = splits[1]
+			}
+		}
+	}
+
+	return handler, nil
+}
+
+func (self *containerdContainerHandler) ContainerReference() (info.ContainerReference, error) {
+	return info.ContainerReference{
+		Id:        self.id,
+		Name:      self.name,
+		Namespace: ContainerdNamespace,
+		Labels:    self.labels,
+		Aliases:   self.aliases,
+	}, nil
+}
+
+func (self *containerdContainerHandler) needNet() bool {
+	// Since containerd does not handle networking ideally we need to return based
+	// on ignoreMetrics list. Here the assumption is the presence of cri-containerd
+	// label
+	if !self.ignoreMetrics.Has(container.NetworkUsageMetrics) {
+		//TODO change it to exported cri-containerd constants
+		return self.labels["io.cri-containerd.kind"] == "sandbox"
+	}
+	return false
+}
+
+func (self *containerdContainerHandler) GetSpec() (info.ContainerSpec, error) {
+	hasFilesystem := !self.ignoreMetrics.Has(container.DiskUsageMetrics)
```

Comment:
What does this mean? I saw both cri-o and docker have this.
However, because we don't collect image filesystem stats, so just make sure.

Reply:
do you collect container disk stats?  If not, you should just set this to false.  I believe this is meant to indicate whether or not this container has disk stats.